README.rst
pytest-flakefinder
===================================

.. image:: https://travis-ci.org/dropbox/pytest-flakefinder.png?branch=master
    :alt: Build Status
    :target: https://travis-ci.org/dropbox/pytest-flakefinder

Runs tests multiple times to expose flakiness.

----

This `Pytest`_ plugin was generated with `Cookiecutter`_ along with `@hackebrot`_'s `Cookiecutter-pytest-plugin`_ template.


Features
--------

* When enabled it will 'multiply' your tests so that they get run multiple times without having to restart pytest.  This helps you find flakiness in your tests.
* You can limit your flake runs to a particular timeout value.


Installation
------------

Install with setup.py::

    python setup.py install

For best flake-finding combine with pytest-xdist::

    pip install pytest-xdist

Usage
-----

Flake Finding
~~~~~~~~~~~~~

Enable plugin for tests::

    pytest --flake-finder

This will run every test the default, 50, times.  Every test is run independently and you can even use xdist to send tests to multiple processes.

To configure the number of runs::

    pytest --flake-finder --flake-runs=runs

To find flakes in one test or a couple of tests you can use pytest's built in test selectiong.

Finding flakes in one test::

    pytest -k test_maybe_flaky --flake-finder

When used with xdist the flake finder can expose many timing related flakes.

Timing Out
~~~~~~~~~~

When using flake-finder as part of a CI run it might be useful to limit the amount of time it will run.

Running with timeout::

    pytest --flake-finder --flake-max-minutes=minutes

Tests started after the timeout will be skipped.

Contributing
------------
Contributions are very welcome. Tests can be run with `tox`_, please ensure
the coverage at least stays the same before you submit a pull request.

Contributors
------------
- euresti_
- nipunn1313_
- goodspark_
- and-semakin_


Issues
------

If you encounter any problems, please `file an issue`_ along with a detailed description.

.. _`Cookiecutter`: https://github.com/audreyr/cookiecutter
.. _`@hackebrot`: https://github.com/hackebrot
.. _`cookiecutter-pytest-plugin`: https://github.com/pytest-dev/cookiecutter-pytest-plugin
.. _`file an issue`: https://github.com/dropbox/pytest-flakefinder/issues
.. _`pytest`: https://github.com/pytest-dev/pytest
.. _`tox`: https://tox.readthedocs.org/en/latest/
.. _euresti: https://github.com/euresti
.. _nipunn1313: https://github.com/nipunn1313
.. _goodspark: https://github.com/goodspark
.. _and-semakin: https://github.com/and-semakin


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [Unreleased]

## [1.0.0] - 2020-07-24
### Deprecated
- Drop support for the deprecated Python 2.7 and 3.4 versions

## [0.1.0] - 2015-10-14
### Added
- Initial Release
- Support for `--flake-finder`, `--flake-runs`, and `--flake-max-minutes` options.
- Unit Tests.


tests/test_flakefinder.py
"""Tests for the flakefinder plugin."""
import pytest
import pytest_flakefinder

def test_help_message(testdir):
    """Test that the options don't change names."""
    result = testdir.runpytest(
        '--help',
    )
    # fnmatch_lines does an assertion internally
    result.stdout.fnmatch_lines([
        'flakefinder:',
        '*--flake-finder*',
        '*--flake-runs=runs*',
        '*--flake-max-minutes=minutes*',
    ])

@pytest.mark.parametrize("flags, runs", [
    ([], pytest_flakefinder.DEFAULT_FLAKE_RUNS),
    (["--flake-runs=1"], 1),
    (["--flake-runs=25"], 25),
])
def test_repeat_success(testdir, flags, runs):
    """Test that pytest-style tests are duplicated correctly."""

    # The test file.
    testdir.makepyfile("""
        def test():
            assert True
    """)

    # Run pytest.
    flags += ['--flake-finder', '-v']
    result = testdir.runpytest(*flags)

    # Check output.
    result.stdout.fnmatch_lines(
        # Wildcard at end because different Python+pytest combos pluralize 'item'
        ['collecting ... collected %d item*' % runs] +
        # fnmatch doesn't like `[` characters so I use `?`.
        ['*::test?%d? PASSED*' % i for i in range(runs)]
    )
    assert result.ret == 0

@pytest.mark.parametrize("flags, runs", [
    ([], pytest_flakefinder.DEFAULT_FLAKE_RUNS),
    (["--flake-runs=1"], 1),
    (["--flake-runs=25"], 25),
])
def test_unittest_repeats(testdir, flags, runs):
    """Test that flake-finder works with unittest.TestCase."""

    # The test file.
    testdir.makepyfile("""
        import unittest

        class TestAwesome(unittest.TestCase):
            def test(self):
                assert True
     """)

    # Run pytest.
    flags += ['--flake-finder', '-v']
    result = testdir.runpytest(*flags)

    # Check output.
    result.stdout.fnmatch_lines(
        # NB: unitest TestCases don't increment the collected items.
        # Wildcard at end because different Python+pytest combos pluralize 'item'
        ['collecting ... collected 1 item*'] +
        ['*::TestAwesome::test PASSED*' for _ in range(runs)]
    )
    assert result.ret == 0

@pytest.mark.parametrize("flags, runs", [
    (["--flake-runs=10"], 10),
    (["--flake-runs=25"], 25),
    ([], pytest_flakefinder.DEFAULT_FLAKE_RUNS),
])
def test_flaky_test(testdir, flags, runs):
    """Test that flaky tests fail the tests."""
    # This test will fail on the 20th call
    testdir.makepyfile("""
        count = [-1]
        def test():
            count[0] += 1
            assert count[0] != 20
    """)

    # Run pytest.
    flags += ['--flake-finder', '-v']
    result = testdir.runpytest(*flags)

    # Check output.
    result.stdout.fnmatch_lines(
        ['collecting ... collected %d items' % runs] +
        ['*::test?%d? PASSED*' % i for i in range(min(runs, 20))] +
        ['*::test?%d? FAILED*' % i for i in range(20, min(runs, 21))] +
        ['*::test?%d? PASSED*' % i for i in range(21, runs)]
    )
    assert result.ret == 0 if runs < 20 else 1

def test_parametrized_tests(testdir):
    """Test that parametrized tests get multiplied correctly too."""
    testdir.makepyfile("""
        import pytest

        @pytest.mark.parametrize("number", (1, 5, 10))
        def test(number):
            assert number
    """)

    # Run pytest.
    flags = ['--flake-finder', '-v']
    result = testdir.runpytest(*flags)

    # Check output.
    result.stdout.fnmatch_lines(
        ['collecting ... collected 150 items'] +
        ['*::test?%d-%d? PASSED*' % (i, x)
         for i in range(pytest_flakefinder.DEFAULT_FLAKE_RUNS)
         for x in [1, 5, 10]]
    )
    assert result.ret == 0

@pytest.mark.parametrize("minutes", (1, 5, 10))
def test_flake_max_minutes(testdir, minutes):
    """Test --flake-max-minutes= option."""
    # On the 10th iteration we monkey-patch time.time to advance time past the
    # limit.
    testdir.makepyfile("""
        import time

        class TestTimePatch(object):

            def setup_class(cls):
                cls.count = 1
                cls.orig_time = time.time

            def teardown_class(cls):
                # Put back the old time when the test is over.  That way we
                # don't report the tests as taking 900 seconds.
                time.time = cls.orig_time

            def test(self):
                if TestTimePatch.count == 10:
                    cur_time = time.time()
                    # Move time 1 second past the limit at least.
                    time.time = lambda: cur_time + %d * 60 + 1
                TestTimePatch.count += 1
    """ % minutes)

    # Run pytest.
    result = testdir.runpytest('--flake-finder', '-v', '--flake-max-minutes=%d' % minutes, '-s')

    passing_runs = 10
    runs = pytest_flakefinder.DEFAULT_FLAKE_RUNS

    # Check output.
    result.stdout.fnmatch_lines(
        # fnmatch doesn't like `[` characters so I use `?`.
        ['collecting ... collected %d items*' % runs] +
        ['*::test?%d? PASSED*' % i for i in range(min(runs, passing_runs))] +
        ['*::test?%d? SKIPPED*' % i for i in range(passing_runs, runs)] +
        # Test for the test, make sure the time isn't modified when coming out.
        ['* 10 passed, 40 skipped in ?.??*s*']
    )
    assert result.ret == 0

def test_flake_derived_classes(testdir):
    """Tests that if two tests share the same function they still get duped properly."""

    testdir.makepyfile("""
    import unittest

    class Base(object):
        def runTest(self):
            pass

    class TestMoreAwesome(Base, unittest.TestCase):
        pass

    class TestAwesome(Base, unittest.TestCase):
        pass
    """)

    # Run pytest.
    flags = ['--flake-finder', '-v', '-k', 'TestAwesome']
    result = testdir.runpytest(*flags)

    # Check output.
    result.stdout.fnmatch_lines(
        # NB: unitest TestCases don't increment the collected items.
        ['collecting ... collected 2 items*'] +
        # The bug was that TestAwesome wouldn't run multiple times.
        ['*::TestAwesome::runTest PASSED*'
         for _ in range(pytest_flakefinder.DEFAULT_FLAKE_RUNS)]
    )
    assert result.ret == 0

def test_fixture_parametrization_with_similar_test_names(testdir):
    """
    Ensure tests with the same name in diff namespaces do not cause overloading of
    flake runs.
    """
    testdir.makepyfile("""
    class TestA(object):
        def test_something(self):
            pass

    class TestB(object):
        def test_something(self):
            pass

    def test_something():
        pass
    """)

    # Run pytest.
    flags = ['--flake-finder', '-v']
    result = testdir.runpytest(*flags)

    # Check output.
    num_runs = pytest_flakefinder.DEFAULT_FLAKE_RUNS
    result.stdout.fnmatch_lines(
        ['collecting ... collected 150 items*'] +
        ['*::test_something?%d? PASSED*' % i for i in range(num_runs)]
    )
    assert result.ret == 0


pytest_flakefinder.py
"""Run a test multiple times to see if it's flaky."""

import copy
import pytest
import time

DEFAULT_FLAKE_RUNS = 50
DEFAULT_FLAKE_MINUTES = 0


def pytest_addoption(parser):
    """Add options for flakefinder plugin."""
    group = parser.getgroup('flakefinder')

    group.addoption("--flake-finder",
                    action='store_true',
                    dest="flake_finder_enable",
                    default=False,
                    help="create multiple copies of all the selected tests.")
    group.addoption("--flake-runs",
                    action='store',
                    dest="flake_runs",
                    default=DEFAULT_FLAKE_RUNS,
                    type=int,
                    metavar="runs",
                    help="number of times to repeat the tests. (default: %default)")
    group.addoption("--flake-max-minutes",
                    action="store",
                    dest="flake_max_minutes",
                    default=DEFAULT_FLAKE_MINUTES,
                    type=int,
                    metavar="minutes",
                    help="Don't run for longer than this parameter. (default: %default)")

def pytest_configure(config):
    """Register the plugin if needed."""
    if config.getoption('flake_finder_enable'):
        config.pluginmanager.register(FlakeFinderPlugin(config))


class FlakeFinderPlugin(object):
    """This is a pytest plugin that multiplies all selected tests by `flake_runs`."""

    def __init__(self, config):
        self.flake_runs = config.getoption('flake_runs')
        self.expires = config.getoption('flake_max_minutes')
        if self.expires:
            self.expires = time.time() + self.expires * 60

    @pytest.mark.tryfirst
    def pytest_generate_tests(self, metafunc):
        """For all true pytest tests use metafunc to add all the duplicates."""
        # This is safer because otherwise test with fixtures might not be setup correctly.
        # Parameterization requires the test function to accept the permutation as
        # either an argument or depend on it as a fixture. Use fixture so the test
        # function signature is not changed. Prefix with underscores and suffix with
        # function name to reduce odds of collision with other fixture names.
        fixture_name = '__flakefinder_{}'.format(metafunc.function.__name__)
        metafunc.fixturenames.append(fixture_name)
        metafunc.parametrize(
            argnames=fixture_name,
            argvalues=list(range(self.flake_runs)),
        )
        metafunc.function._pytest_duplicated = True

    @pytest.mark.tryfirst
    def pytest_collection_modifyitems(self, items):
        """Add unitest tests to the collection."""
        # Some tests (e.g. unittest.TestCase) don't pass through
        # pytest_generate_tests. For those we use the old "add them multiple
        # times to the items list" trick.
        #
        # Also we want to @tryfirst so that we go before randomizing the list.
        for item in list(items):
            if not getattr(item.function, '_pytest_duplicated', None):
                for i in range(self.flake_runs - 1):
                    cpy = copy.copy(item)
                    # HAX
                    # Ensure initialization for the copied request works for _pytest.TestCaseFunction
                    # without this, funcargs ends up being None
                    cpy._initrequest()
                    items.append(cpy)

    def pytest_runtest_call(self, item):
        """Skip tests if we've run out of time."""
        if self.expires and self.expires < time.time():
            pytest.skip("time bound exceeded")
